pkg/ingester/instance.go

```diff
@@ -211,6 +221,73 @@ func (i *instance) Push(ctx context.Context, req *logproto.PushRequest) error {
 	return appendErr
 }
 
+// mountGRPCError mounts an error to be returned in a GRPC call.
+//
+// The returned error is enriched with the gRPC error status and with a list of details.
+// As of now, the list of details is a list of all streams that were rate-limited. This list can be used
+//   by the distributor to fine tune streams that were limited.
+func mountGRPCError(s *stream, failedEntriesWithError []entryWithError, totalEntries int) error {
+	if len(failedEntriesWithError) == 0 {
+		return nil
+	}
+
+	lastEntryWithErr := failedEntriesWithError[len(failedEntriesWithError)-1]
+	_, ok := lastEntryWithErr.e.(*validation.ErrStreamRateLimit)
+	outOfOrder := chunkenc.IsOutOfOrderErr(lastEntryWithErr.e)
+	if !outOfOrder && !ok {
+		return lastEntryWithErr.e
+	}
+
+	var statusCode int
+	if outOfOrder {
+		statusCode = http.StatusBadRequest
+	}
+	if ok {
+		// per-stream or ingestion limited.
+		statusCode = http.StatusTooManyRequests
+	}
+
+	// Return a http status 4xx request response with all failed entries.
+	buf := bytes.Buffer{}
+	streamName := s.labelsString
+
+	limitedFailedEntries := failedEntriesWithError
+	if maxIgnore := s.cfg.MaxReturnedErrors; maxIgnore > 0 && len(limitedFailedEntries) > maxIgnore {
+		limitedFailedEntries = limitedFailedEntries[:maxIgnore]
+	}
+
+	hadPerStreamError := false
+	for _, entryWithError := range limitedFailedEntries {
+		if !hadPerStreamError {
```

Comment:
We don't need to check for `perStreamError` here because we already have the status code `http.StatusTooManyRequests`

Reply:
But we also use `http.StatusTooManyRequests` for the total ingestion rate limit (which is different from the per-stream rate limit). If we mark a stream as rate-limited when it actually did hit the total ingestion rate, we might end up wrongly sharding a stream that shouldn't.

WDYT?

Reply:
The errors are checked in different places. The distributor returns 429 for global rate limiting so the only 429s we'll ever get from the ingester are for per stream rate limiting.